src/pybind/cephfs/cephfs.pyx

```diff
@@ -3013,16 +3047,20 @@ class RmtreeDir:
     helper for class NonRecursiveRmtree.
     '''
 
-    def __init__(self, fs, path):
-        self.fs = fs
+    def __init__(self, rmtree_run_obj, rel_path):
+        self.fs = rmtree_run_obj.fs
+        self.st_root_fd = rmtree_run_obj.st_root_fd
+        self.trash_path = rmtree_run_obj.trash_path
+        self.rel_path = rel_path
+        self.abs_path = os.path.join(self.trash_path, self.rel_path)
+
+        self.name = os.path.basename(self.rel_path)
+        if isinstance(self.name, str):
+            self.name = self.name.encode('utf-8')
 
-        self.path = path
-        if isinstance(self.path, str):
-            self.path = self.path.encode('utf-8')
-        self.name = os.path.basename(self.path)
         # XXX: exception (if) raised here should be handled by caller based on
         # the context.
-        self.handle = self.fs.opendir(self.path)
+        self.handle = self.fs.opendir(self.abs_path)
```

Comment:
You should no longer need the `abspath` any more. This should be converted to:

`self.fs.fdopendir(dirfd)`

You need a python binding for that too.

(If you need the abspath anywhere in this code, you probably are missing a binding to work with file descriptors (and relative paths).)